pdd/PDD-multistream.md

```diff
@@ -0,0 +1,79 @@
+# PDD for MultiStream (example/use case)
+
+## Protocol
+
+Defined here:
+- https://github.com/ipfs/specs/blob/wire/protocol/network/wire.md#multistream---self-describing-protocol-stream
+- https://github.com/ipfs/specs/blob/wire/protocol/network/wire.md#multistream-selector---self-describing-protocol-stream-selector
+
+## Protocol Compliance Tests Spec
+
+Given the protocol spec, an implementation of the multistream-select protocol has to comply with the following scenarios:
+
+#### 1 - push-stream
+
+In a push-stream example (one-way stream), we have two agents:
+
+- 'broadcast' - where messages are emited 
+- 'silent' - listener to the messages emited by the broadcast counterparty
+
+Compliance test 1 (human readable format):
+```
+# With a connection established between silent - broadcast
+< /multistream/1.0.0       # multistream header
+< /bird/3.2.1              # Protocol header
+< hey, how is it going?    # First protocol message
+```
+
+#### 2 - duplex-stream
+
+In a duplex-stream example (interactive conversation), we have two agents:
+
+- 'select' - waiting for connections, hosts several protocols from where a client can pick from
+- 'interactive' - connects to a select agent and queries that agent for a specific protocol
+
+Compliance test 2 (human readable format):
+```
+# With a connection established between interactive - select
+< /multistream/1.0.0
+> /multistream/1.0.0
+> ls
+< ["/dogs/0.1.0","/cats/1.2.11"]
+> /mouse/1.1.0
+< na
+> /dogs/0.1.0
+< /dogs/0.1.0
+> hey
+```
```

Comment:
ideally the test could try to fuzz all possible message orderings. e.g.:

```
# order 1
< /multistream/1.0.0
> /multistream/1.0.0
> ls
< ["/dogs/0.1.0","/cats/1.2.11"]
> /mouse/1.1.0
< na
> /dogs/0.1.0
< /dogs/0.1.0
> hey

# order 2
> /multistream/1.0.0
< /multistream/1.0.0
> ls
< ["/dogs/0.1.0","/cats/1.2.11"]
> /mouse/1.1.0
< na
> /dogs/0.1.0
< /dogs/0.1.0
> hey

# order 3
> /multistream/1.0.0
> ls
< /multistream/1.0.0
< ["/dogs/0.1.0","/cats/1.2.11"]
> /mouse/1.1.0
< na
> /dogs/0.1.0
< /dogs/0.1.0
> hey

... and so on down ...

# another (this one is weird from an "interactivity" perspective
# but --for correctness-- it should be tested to ensure things behave
# as expected).
> /multistream/1.0.0
> ls
> /mouse/1.1.0
> /dogs/0.1.0
> hey
< /multistream/1.0.0
< ["/dogs/0.1.0","/cats/1.2.11"]
< na
< /dogs/0.1.0
```

not sure if this is worth worrying about or not